server/automute_channel.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/plugin"
+
+	"github.com/pkg/errors"
+)
+
+func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, actor *model.User) {
+	_, _ = p.updateAutomutingOnUserJoinedChannel(c, channelMember.UserId, channelMember.ChannelId)
+}
+
+func (p *Plugin) updateAutomutingOnUserJoinedChannel(c *plugin.Context, userID string, channelID string) (bool, error) {
+	if automuteEnabled := p.getAutomuteIsEnabledForUser(userID); !automuteEnabled {
+		return false, nil
+	}
+
+	if canAutomute, err := p.isChannelLinked(channelID); err != nil {
+		p.API.LogError(
+			"Unable to check if channel is linked to update automuting when a user has joined the channel",
+			"UserID", userID,
+			"ChannelID", channelID,
+			"Error", err.Error(),
+		)
+		return false, errors.Wrap(err, "Unable to update automuting when a user has joined a channel")
+	} else if !canAutomute {
+		// Only automute channels that are linked
+		return false, nil
+	}
+
+	err := p.setChannelMembersAutomuted([]*model.ChannelMemberIdentifier{{UserId: userID, ChannelId: channelID}}, true)
+	return err == nil, err
+}
+
+func (p *Plugin) ChannelHasBeenCreated(c *plugin.Context, channel *model.Channel) {
+	_ = p.updateAutomutingOnChannelCreated(channel)
+}
+
+func (p *Plugin) updateAutomutingOnChannelCreated(channel *model.Channel) error {
+	if !channel.IsGroupOrDirect() {
+		// Assume that newly created channels can never be linked by the time this is called
+		return nil
+	}
+
+	return p.updateAutomutingForChannelMembers(channel.Id, true)
+}
+
+func (p *Plugin) updateAutomutingOnChannelLinked(channelID string) error {
+	// This simply mutes the channel for all users with automuting enabled, regardless of their settings before. It
+	// doesn't pay attention to if the user manually muted the channel beforehand.
+	return p.updateAutomutingForChannelMembers(channelID, true)
+}
+
+func (p *Plugin) updateAutomutingOnChannelUnlinked(channelID string) error {
+	// This simply unmutes the channel for all users with automuting enabled, regardless of their settings before. It
+	// doesn't pay attention to if the user manually muted the channel beforehand to keep it muted.
+	return p.updateAutomutingForChannelMembers(channelID, false)
```

Comment:
This is unfortunate, but not blocking. It would be really neat if we could somehow distinguish the two types of muting, maybe by remembering the timestamp for manual vs. auto events.

Reply:
I could save the previous state in the `NotifyProps` and then restore it when the channel is no longer automuted, but I honestly don't think that will come up often enough to bother doing that. I'd assume channels won't go back and forth between connected often and that, if they are unlinked, it'll be long enough that the user wouldn't remember if they had it previously muted